iceoryx_dust/vocabulary/include/iox/detail/span_iterator.hpp

```diff
@@ -0,0 +1,223 @@
+    {
+    }
+
+    explicit constexpr operator span_iterator<const T>() const noexcept
+    {
+        return {m_begin, m_end, m_current};
+    }
+
+    constexpr reference operator*() const noexcept
+    {
+        assert(m_begin && m_end);
+        assert(m_begin <= m_current && m_current < m_end);
+        return *m_current;
+    }
+
+    constexpr pointer operator->() const noexcept
+    {
+        assert(m_begin && m_end);
+        assert(m_begin <= m_current && m_current < m_end);
+        return m_current;
+    }
+    constexpr span_iterator& operator++() noexcept
+    {
+        assert(m_begin && m_current && m_end);
+        assert(m_current < m_end);
+
+        ++m_current;
+        return *this;
+    }
+
+    // Rule DCL21-CPP is deprecated
+    // NOLINTNEXTLINE(cert-dcl21-cpp)
+    constexpr span_iterator operator++(int) noexcept
+    {
+        span_iterator ret = *this;
+        ++*this;
+        return ret;
+    }
+
+    constexpr span_iterator& operator--() noexcept
+    {
+        assert(m_begin && m_end);
+        assert(m_begin < m_current);
+        --m_current;
+        return *this;
+    }
+
+    // Rule DCL21-CPP is deprecated
+    // NOLINTNEXTLINE(cert-dcl21-cpp)
+    constexpr span_iterator operator--(int) noexcept
+    {
+        span_iterator ret = *this;
+        --*this;
+        return ret;
+    }
+
+    constexpr span_iterator& operator+=(const difference_type n) noexcept
+    {
+        if (n != 0)
+        {
+            assert(m_begin && m_current && m_end);
+        }
+        if (n > 0)
+        {
+            assert(m_end - m_current >= n);
+        }
+        if (n < 0)
+        {
+            assert(m_current - m_begin >= -n);
+        }
+
+        m_current += n;
+        return *this;
+    }
+
+    constexpr span_iterator operator+(const difference_type n) const noexcept
+    {
+        span_iterator ret = *this;
+        ret += n;
+        return ret;
+    }
+
+    friend constexpr span_iterator operator+(const difference_type n, const span_iterator& rhs) noexcept
+    {
+        return rhs + n;
+    }
+
+    constexpr span_iterator& operator-=(const difference_type n) noexcept
+    {
+        if (n != 0)
+        {
+            assert(m_begin && m_current && m_end);
+        }
+        if (n > 0)
+        {
+            assert(m_current - m_begin >= n);
+        }
+        if (n < 0)
+        {
+            assert(m_end - m_current >= -n);
+        }
+        m_current -= n;
+        return *this;
+    }
+
+    constexpr span_iterator operator-(const difference_type n) const noexcept
+    {
+        span_iterator ret = *this;
+        ret -= n;
+        return ret;
+    }
+
+    template <class Type2, std::enable_if_t<std::is_same<std::remove_cv_t<Type2>, value_type>::value, int> = 0>
+    constexpr difference_type operator-(const span_iterator<Type2>& rhs) const noexcept
+    {
+        assert(m_begin == rhs.m_begin && m_end == rhs.m_end);
+        return m_current - rhs.m_current;
```

Comment:
Is it possible that `m_current` is still a `nullptr` at this time?

Reply:
@elBoberido The first `assert` checks `m_begin` and `m_end` in both `span_iterator`s are equal. That means both `span_iterator`s are constructed with the same `span`. Hence, `m_current` can't be `nullptr` because the `span` is not empty, in such a case also `m_begin` and `m_end` would both be `nullptr`.